### PR TITLE
Plug in the new routing introduced in #5046

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
- Plug in the new routing for the new added routing features
- Re-implement time boundary refresh as a general broker segment refresh to handle the update of time info, partition info etc.
- Add PinotBrokerRouting resource to provide APIs to manage the broker routing
  - PUT `/routing/{tableName}`: Build/rebuild the routing for a table
  - PUT `/routing/refresh/{tableName}/{segmentName}`: Refresh the routing for a segment
  - DELETE `/routing/{tableName}`: Remove the routing for a table
- Add a new debug API to help test the segment pruning
  - GET `/debug/routingTable?pql={pql}`: Get the routing table for a query

(Backward-incompatible) Broker rest API changes:
- GET `/debug/timeBoundary/{tableName}` will return NOT_FOUND status when time boundary does not exist
- GET `/debug/routingTable/{tableName}` will return the actual routing table instead of the snapshots because snapshot is no longer available for the new routing
- GET `/health` will return the status from the ServiceStatus in order to reflect the real status of the broker

(Backward-incompatible) Time boundary refresh:
Time boundary refresh is merged into the general broker segment refresh and is triggered by the SegmentRefreshMessage.
In case controller and broker are running in different versions, it won't have any effect.
Since time boundary refresh is needed in very rare scenario, it should have very little impact.

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/messages/SegmentRefreshMessage.java [hunk 1]
```diff
@@ -19,52 +19,60 @@
...
 
 
 /**
- * This (helix) message is sent from the controller to the server when a request is received to refresh
+ * This (Helix) message is sent from the controller to brokers and servers when a request is received to refresh
  * an existing segment.
  *
- * There is one mandatory field in the message -- the CRC of the new segment.
+ * NOTE: Changing this class to include new fields is a change in the protocol, so the new fields must be made optional,
+ * and coded in such a way that either controller, broker or server may be upgraded first.
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/messages/SegmentRefreshMessage.java]: Can we update the javadoc to state the time boundary refresh is now done as part of SegmentRefreshMessage?

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/helix/HelixBrokerStarter.java [hunk 2]
```diff
@@ -82,16 +82,16 @@
   private ZkHelixPropertyStore<ZNRecord> _propertyStore;
   private HelixDataAccessor _helixDataAccessor;
 
-  // Cluster change handlers
-  private HelixExternalViewBasedRouting _helixExternalViewBasedRouting;
```
Comment (siddharthteotia) [inline pinot-broker/src/main/java/org/apache/pinot/broker/broker/helix/HelixBrokerStarter.java]: It looks like, with these changes we can HelixExternalViewBasedRouting class and RoutingTable interface? (assuming RoutingManageris backward compatible)

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/helix/BrokerUserDefineMessageHandlerFactory.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import org.apache.helix.NotificationContext;
+import org.apache.helix.messaging.handling.HelixTaskResult;
+import org.apache.helix.messaging.handling.MessageHandler;
+import org.apache.helix.messaging.handling.MessageHandlerFactory;
+import org.apache.helix.model.Message;
+import org.apache.pinot.broker.routing.v2.RoutingManager;
+import org.apache.pinot.common.messages.SegmentRefreshMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
```
Comment (siddharthteotia) [inline pinot-broker/src/main/java/org/apache/pinot/broker/broker/helix/BrokerUserDefineMessageHandlerFactory.java]: (nit): Can we please add Javadoc?

File: pinot-common/src/main/java/org/apache/pinot/common/messages/SegmentRefreshMessage.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/messages/SegmentRefreshMessage.java]: Added

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/helix/HelixBrokerStarter.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-broker/src/main/java/org/apache/pinot/broker/broker/helix/HelixBrokerStarter.java]: Yes. The code cleanup will be in the following pr.

File: pinot-broker/src/main/java/org/apache/pinot/broker/broker/helix/BrokerUserDefineMessageHandlerFactory.java (hunk 3 above)
Comment (Jackie-Jiang) [reply pinot-broker/src/main/java/org/apache/pinot/broker/broker/helix/BrokerUserDefineMessageHandlerFactory.java]: Added

File: pinot-broker/src/main/java/org/apache/pinot/broker/api/resources/PinotBrokerDebug.java [hunk 4]
```diff
@@ -23,77 +23,96 @@
...
+import org.apache.pinot.broker.routing.v2.timeboundary.TimeBoundaryInfo;
 import org.apache.pinot.common.config.TableNameBuilder;
-import org.apache.pinot.common.metrics.BrokerMeter;
-import org.apache.pinot.common.metrics.BrokerMetrics;
-import org.apache.pinot.common.utils.CommonConstants;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import org.apache.pinot.common.utils.CommonConstants.Helix.TableType;
+import org.apache.pinot.core.transport.ServerInstance;
+import org.apache.pinot.pql.parsers.Pql2Compiler;
```
Comment (siddharthteotia) [inline pinot-broker/src/main/java/org/apache/pinot/broker/api/resources/PinotBrokerDebug.java]: I generally like the REST API resources to document in the javadoc each API endpoint and purpose. But upto you.

**Pull request merged**